### PR TITLE
Allow to explicitly specify feature qualifier per json file

### PR DESCRIPTION
**Pull request by shchukax**
This PR addresses issue #933 - to allow explicitly specifying qualifiers for each json.

### CODE REVIEW

File: src/main/java/net/masterthought/cucumber/Configuration.java [hunk 1]
```diff
@@ -310,4 +307,20 @@ public void setNotFailingStatuses(Set<Status> notFailingStatuses) {
...
+
+    public void setQualifier(@NonNull String jsonFile, @NonNull String qualifier) {
+        qualifiers.put(jsonFile, qualifier);
+    }
+
+    public String getQualifier(@NonNull String jsonFile) {
+        return qualifiers.get(jsonFile);
+    }
+
+    public boolean isQualifierSet(@NonNull String jsonFile) {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/Configuration.java:319]: containsQualifier?
hasQualifier?

File: src/main/java/net/masterthought/cucumber/ReportParser.java [hunk 2]
```diff
@@ -93,7 +93,9 @@ public ReportParser(Configuration configuration) {
                 LOG.log(Level.INFO, "File '{0}' does not contain features", jsonFile);
             }
             String jsonFileName = extractQualifier(jsonFile);
-            Arrays.stream(features).forEach(feature -> feature.setQualifier(jsonFileName));
+            Arrays.stream(features).forEach(feature ->
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/ReportParser.java:96]: did you test this code as well?

File: src/test/java/net/masterthought/cucumber/ConfigurationTest.java [hunk 3]
```diff
@@ -186,6 +186,90 @@ public void getDirectorySuffixWithSeparatorForEmptySuffix_ReturnsEmptyString() {
...
+        Configuration configuration = new Configuration(outputDirectory, projectName);
+
+        // when
+        configuration.setQualifier(jsonFile, qualifier);
+        configuration.removeQualifier(jsonFile);
+
+        // then
+        assertThat(configuration.isQualifierSet(jsonFile)).isFalse();
+    }
+
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/ConfigurationTest.java:272]: I like that above - great to have it.

I need one more to test with https://github.com/damianszczepanik/cucumber-reporting/blob/master/src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java#L149 that the effects is present in overview page as well. Copy mentioned test, set qualifier and examine results

File: src/main/java/net/masterthought/cucumber/Configuration.java (hunk 1 above)
Comment (shchukax) [reply src/main/java/net/masterthought/cucumber/Configuration.java]: Renamed to `containsQualifier`

File: src/main/java/net/masterthought/cucumber/ReportParser.java (hunk 2 above)
Comment (shchukax) [reply src/main/java/net/masterthought/cucumber/ReportParser.java:96]: I tested manually - but now added an integration test that covers this.

File: src/test/java/net/masterthought/cucumber/ConfigurationTest.java (hunk 3 above)
Comment (shchukax) [reply src/test/java/net/masterthought/cucumber/ConfigurationTest.java:272]: Added an integration test `FeaturesOverviewPageIntegrationTest.generatePage_WithSpecifiedQualifier_generatesStatsTableBody` checking that the qualifier has been set and used correctly.

File: src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java [hunk 4]
```diff
@@ -149,6 +150,38 @@ public void generatePage_generatesStatsTableBody() {
         secondRow.getReportLink().hasLabelAndAddress("Second feature", "report-feature_1_1515379431.html");
     }
 
+    @Test
+    public void generatePage_WithSpecifiedQualifier_generatesStatsTableBody() {
+
+        // given
+        String testQualifier = "testQualifier";
+        configuration.setQualifier(FilenameUtils.removeExtension(SAMPLE_JSON), testQualifier);
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java:158]: I would expect https://github.com/damianszczepanik/cucumber-reporting/pull/936/files#diff-87720169362dbe935db6e3bfbb83b6a6R311 takes filename WITH extension while here looks like without

File: src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java [hunk 5]
```diff
@@ -149,6 +150,38 @@ public void generatePage_generatesStatsTableBody() {
...
+
+        // then
+        DocumentAssertion document = documentFrom(page.getWebPage());
+        TableRowAssertion[] bodyRows = document.getReport().getTableStats().getBodyRows();
+
+        assertThat(bodyRows).hasSize(2);
+
+        TableRowAssertion firstRow = bodyRows[0];
+        firstRow.hasExactValues("1st feature", testQualifier, "10", "0", "0", "0", "0", "10", "1", "0", "1", "1:39.263", "Passed");
+        firstRow.hasExactCSSClasses("tagname", "", "passed", "", "", "", "", "total", "passed", "", "total", "duration", "passed");
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java:174]: remove from this line to end of section and same for second feature - this allows to focus what is really tested here

File: src/main/java/net/masterthought/cucumber/Configuration.java [hunk 6]
```diff
@@ -310,4 +307,20 @@ public void setNotFailingStatuses(Set<Status> notFailingStatuses) {
             this.notFailingStatuses = notFailingStatuses;
         }
     }
+
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/Configuration.java:310]: add javadoc here

File: src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java (hunk 4 above)
Comment (shchukax) [reply src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java:158]: No, the filename is taken without the extension.  When the qualifier is retrieved in `ReportParser.java` lines 95-97, it uses the result of `extractQualifier(jsonFile)` as the key to get the qulifier - and that method strips out the extension.

File: src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java (hunk 5 above)
Comment (shchukax) [reply src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java:174]: Done

File: src/main/java/net/masterthought/cucumber/Configuration.java (hunk 6 above)
Comment (shchukax) [reply src/main/java/net/masterthought/cucumber/Configuration.java:310]: Done

File: src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java (hunk 4 above)
Comment (shchukax) [reply src/test/java/net/masterthought/cucumber/generators/integrations/FeaturesOverviewPageIntegrationTest.java:158]: No, filename is expected to be without the extension, as can be seen in `ReportParser.java` lines 95-97.  Results of call to `extractQualifier(jsonFile)` is used as the key for lookup and that method strips `.json` extension.

File: src/main/java/net/masterthought/cucumber/Configuration.java (hunk 6 above)
Comment (shchukax) [reply src/main/java/net/masterthought/cucumber/Configuration.java:310]: Done

Comment (damianszczepanik) [reply src/main/java/net/masterthought/cucumber/Configuration.java:310]: it is not done

File: src/main/java/net/masterthought/cucumber/Configuration.java [hunk 7]
```diff
@@ -310,4 +307,20 @@ public void setNotFailingStatuses(Set<Status> notFailingStatuses) {
...
+    public void setQualifier(@NonNull String jsonFile, @NonNull String qualifier) {
+        qualifiers.put(jsonFile, qualifier);
+    }
+
+    public String getQualifier(@NonNull String jsonFile) {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/Configuration.java:315]: `jsonFile` -> `jsonFileName` to make it clear it does not contain extension?

like `private String extractQualifier(String jsonFileName)`

File: src/main/java/net/masterthought/cucumber/Configuration.java [hunk 8]
```diff
@@ -310,4 +307,20 @@ public void setNotFailingStatuses(Set<Status> notFailingStatuses) {
             this.notFailingStatuses = notFailingStatuses;
         }
     }
+
+    public void setQualifier(@NonNull String jsonFile, @NonNull String qualifier) {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/Configuration.java:311]: can you update about it  README.md and LiveDemoTest - very basic sample is good enough

Comment (shchukax) [reply src/main/java/net/masterthought/cucumber/Configuration.java:311]: @damianszczepanik Should I amend existing LiveDemoTest or add an additional one?  Similarly with README.md - amend existing example or add another one?

Comment (damianszczepanik) [reply src/main/java/net/masterthought/cucumber/Configuration.java:311]: amend :) both!

File: src/main/java/net/masterthought/cucumber/Configuration.java (hunk 7 above)
Comment (shchukax) [reply src/main/java/net/masterthought/cucumber/Configuration.java]: Done

File: src/main/java/net/masterthought/cucumber/Configuration.java (hunk 8 above)
Comment (shchukax) [reply src/main/java/net/masterthought/cucumber/Configuration.java]: Done

**Pull request merged**